cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java

```diff
@@ -552,6 +566,345 @@ public void testDeletedProgramHandlerStage() throws Exception {
+        .build(),
+      getProperties(app, MetadataScope.SYSTEM));
+    Assert.assertEquals(ImmutableSet.of(AllProgramsApp.class.getSimpleName(), AllProgramsApp.NAME),
+                        getTags(app, MetadataScope.SYSTEM));
+    // verify program system metadata
+    assertProgramSystemMetadata(Id.Program.from(app, ProgramType.FLOW, AllProgramsApp.NoOpFlow.NAME), "Realtime");
+    assertProgramSystemMetadata(Id.Program.from(app, ProgramType.WORKER, AllProgramsApp.NoOpWorker.NAME), "Realtime");
+    assertProgramSystemMetadata(Id.Program.from(app, ProgramType.SERVICE, AllProgramsApp.NoOpService.NAME), "Realtime");
+    assertProgramSystemMetadata(Id.Program.from(app, ProgramType.MAPREDUCE, AllProgramsApp.NoOpMR.NAME), "Batch");
+    assertProgramSystemMetadata(Id.Program.from(app, ProgramType.SPARK, AllProgramsApp.NoOpSpark.NAME), "Batch");
+    assertProgramSystemMetadata(Id.Program.from(app, ProgramType.WORKFLOW, AllProgramsApp.NoOpWorkflow.NAME), "Batch");
+  }
+
+  @Test
+  public void testSearchUsingSystemMetadata() throws Exception {
+    deploy(AllProgramsApp.class);
+    // search artifacts
+    assertArtifactSearch();
+    // search app
+    Id.Application app = Id.Application.from(Id.Namespace.DEFAULT, AllProgramsApp.NAME);
+    assertAppSearch(app);
+    // search programs
+    assertProgramSearch(app);
+    // search data entities
+    assertDataEntitySearch();
+  }
+
+  private void assertProgramSystemMetadata(Id.Program programId, String mode) throws Exception {
+    Assert.assertTrue(getProperties(programId, MetadataScope.SYSTEM).isEmpty());
+    Set<String> expected = ImmutableSet.of(programId.getId(), programId.getType().getPrettyName(), mode);
+    if (ProgramType.WORKFLOW == programId.getType()) {
+      expected = ImmutableSet.of(programId.getId(), programId.getType().getPrettyName(), mode,
+                                 AllProgramsApp.NoOpAction.class.getSimpleName(), AllProgramsApp.NoOpMR.NAME);
+    }
+    Assert.assertEquals(expected, getTags(programId, MetadataScope.SYSTEM));
+  }
+
+  private void assertArtifactSearch() throws Exception {
+    // add a plugin artifact.
+    Manifest manifest = new Manifest();
+    manifest.getMainAttributes().put(ManifestFields.EXPORT_PACKAGE,
+                                     AllProgramsApp.AppPlugin.class.getPackage().getName());
+    Id.Artifact pluginArtifact = Id.Artifact.from(Id.Namespace.DEFAULT, "plugins", "1.0.0");
+    Assert.assertEquals(HttpResponseStatus.OK.getCode(),
+                        addPluginArtifact(pluginArtifact, AllProgramsApp.AppPlugin.class, manifest,
+                                          new HashSet<ArtifactRange>()).getStatusLine().getStatusCode());
+    // search using artifact name
+    Set<MetadataSearchResultRecord> expected = ImmutableSet.of(new MetadataSearchResultRecord(pluginArtifact));
+    Set<MetadataSearchResultRecord> results = searchMetadata(Id.Namespace.DEFAULT, "plugins", null);
+    Assert.assertEquals(expected, results);
```

Comment:
For plugin or Schema can we add a metadata with same key as name and make to test
one without plugin: which gives 2 results
another with plugin: which gives 1 to  make sure the scoping works ?